pkg/tfcli/client.go

```diff
@@ -0,0 +1,340 @@
+		return err
+	}
+	if err := c.addPidState(pid); err != nil {
+		return err
+	}
+	c.pInfo.LogStdout()
+	c.pInfo.LogStderr()
+	go func() {
+		logger := c.logger.WithValues("args", args, "executable", pathTerraform, "cwd", c.wsPath)
+		logger.Info("Waiting for process to terminate gracefully.", "arg-from-process", c.pInfo.GetCmd().String())
+		err := c.pInfo.WaitError()
+		stderr, _ := c.pInfo.StderrAsString()
+		stdout, _ := c.pInfo.StdoutAsString()
+		logger = logger.WithValues("stderr", stderr, "stdout", stdout)
+
+		if err != nil {
+			logger.Info("Failed to run Terraform CLI", "error", err)
+		} else {
+			logger.Info("Successfully executed Terraform CLI", "stdout", stdout)
+		}
+		if storeResult != nil {
+			storePath := filepath.Join(c.wsPath, fileStore)
+			if err := storeResult(c, stdout, stderr); err != nil {
+				logger.Info("Failed to store result from Terraform CLI", "store", storePath)
+				return
+			}
+			logger.Info("Successfully stored result from Terraform CLI", "store", storePath)
+		}
+	}()
+	return nil
+}
+
+func (c *Client) syncPipeline(ctx context.Context, ignoreExitErr bool, command string, args ...string) error {
+	var err error
+	if c.pInfo, err = process.New(command, args, c.wsPath, false, true, false, c.logger); err != nil {
+		return errors.Wrapf(err, fmtErrSyncRun, command, args, c.wsPath)
+	}
+	exitErr := &exec.ExitError{}
+	if err = c.pInfo.Run(ctx); err != nil && (!ignoreExitErr || !errors.As(err, &exitErr)) {
+		return errors.Wrapf(err, fmtErrSyncRun, command, args, c.wsPath)
+	}
+	return nil
+}
+
+// load Terraform state into Client's cache
+func (c *Client) loadStateFromWorkspace() error {
+	pathState := filepath.Join(c.wsPath, fileState)
+	var err error
+	c.tfState, err = c.readFile(pathState)
+	if err != nil {
+		return errors.Wrapf(err, fmtErrLoadState, pathState)
+	}
+	return nil
+}
+
+func (c *Client) storeStateInWorkspace() error {
+	if c.tfState == nil {
+		return nil
+	}
+
+	pathState := filepath.Join(c.wsPath, fileState)
+	return errors.Wrapf(c.writeFile(pathState, c.tfState, 0600), fmtErrStoreState, pathState)
+}
+
+func (c *Client) pathExists(path string, isDir bool) (bool, error) {
+	fInfo, err := c.fs.Stat(path)
+	if err == nil && fInfo.IsDir() == isDir {
+		// path exists and is of expected type
+		return true, nil
+	}
+	if err == nil && fInfo.IsDir() != isDir {
+		return false, errors.Errorf(fmtErrPath, path, isDir, fInfo.IsDir())
+	}
+	if err != nil && !errors.Is(err, os.ErrNotExist) {
+		return false, errors.Wrapf(err, fmtErrPath, path, isDir, fInfo.IsDir())
+	}
+	return false, nil
+}
+
+// Close releases resources allocated for this Client.
+// After a call to Close, do not reuse the same handle.
+func (c *Client) Close(_ context.Context) error {
+	var result error
+	if c.pInfo != nil {
+		result = multierr.Combine(errors.Wrap(c.pInfo.Kill(), errDestroyProcess))
+	}
+	if c.wsPath != "" {
+		result = multierr.Combine(result, errors.Wrap(c.fs.RemoveAll(c.wsPath), errDestroyWorkspace))
+	}
```

Comment:
```suggestion
	if c.wsPath != "" {
		if err := c.fs.RemoveAll(c.wsPath); err != nil {
			return errors.Wrap(err, errDestroyWorkspace)
		}
	}
```

It makes it much easier to skim through the code quickly if we use existing idiomatic way of error handling. In fact, I'd suggest not using `multierr.Combine` at all and go for subsequent `if err := ..`s if we need to ahve a series of steps.

Reply:
Here we are tearing down multiple resources and if any of them fails, we would still like to tear the rest down and report any failures to the caller.